src/Microsoft.ML.Data/Transforms/ValueMappingTransformer.cs

```diff
@@ -832,25 +832,52 @@ public override void Train(IHostEnvironment env, RowCursor cursor)
                 }
             }
 
-            public override Delegate GetGetter(Row input, int index)
+            private TValue MapValue(TKey key)
             {
-                var src = default(TKey);
-                ValueGetter<TKey> getSrc = input.GetGetter<TKey>(index);
-                ValueGetter<TValue> retVal =
-                (ref TValue dst) =>
+                if (_mapping.ContainsKey(key))
                 {
-                    getSrc(ref src);
-                    if (_mapping.ContainsKey(src))
-                    {
-                        if (ValueType is VectorType vectorType)
-                            dst = Utils.MarshalInvoke(GetVector<int>, vectorType.ItemType.RawType, _mapping[src]);
-                        else
-                            dst = Utils.MarshalInvoke(GetValue<int>, ValueType.RawType, _mapping[src]);
-                    }
+                    if (ValueType is VectorType vectorType)
+                        return Utils.MarshalInvoke(GetVector<int>, vectorType.ItemType.RawType, _mapping[key]);
                     else
-                        dst = _missingValue;
-                };
-                return retVal;
+                        return Utils.MarshalInvoke(GetValue<int>, ValueType.RawType, _mapping[key]);
+                }
+                else
+                    return _missingValue;
+            }
+
+            public override Delegate GetGetter(Row input, int index)
+            {
+                if (input.Schema[index].Type is VectorType)
+                {
+                    var src = default(VBuffer<TKey>);
+                    var getSrc = input.GetGetter<VBuffer<TKey>>(index);
+
+                    ValueGetter<VBuffer<TValue>> retVal =
+                        (ref VBuffer<TValue> dst) =>
+                        {
+                            getSrc(ref src);
+                            var editor = VBufferEditor.Create(ref dst, src.Length);
+                            var values = src.GetValues();
```

Comment:
>var values = src.GetValues(); [](start = 28, length = 29)

Could `src` be a sparse vector? If so, it would be better to iterate over the indices instead of the values.

Reply:
No need to iterate over the indices I think. Just need to copy indices line 861.

---
In reply to: [249201851](https://github.com/dotnet/machinelearning/pull/2190#discussion_r249201851) [](ancestors = 249201851)

Reply:
You are right GetValues() already gives an array of values that are explicitly represented

---
In reply to: [249225682](https://github.com/dotnet/machinelearning/pull/2190#discussion_r249225682) [](ancestors = 249225682,249201851)